Source/AwsCommonRuntimeKit/sdkutils/CRTAWSEndpointsRequestContext.swift

```diff
@@ -0,0 +1,55 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import AwsCSdkUtils
+
+/// Request context used for resolving endpoint
+public class CRTAWSEndpointsRequestContext {
+    let rawValue: OpaquePointer
+
+    /// Initialize a new request context
+    /// - Parameter allocator: Allocator to use for request context creation
+    public init(allocator: Allocator = defaultAllocator) throws {
+        guard let rawValue = aws_endpoints_request_context_new(allocator.rawValue) else {
+            throw CRTError.awsError(AWSCommonRuntimeError())
+        }
+
+        self.rawValue = rawValue
+    }
+
+    /// Add a string endpoint parameter to the request context
+    /// - Parameters:
+    ///   - name: The name of the parameter
+    ///   - value: The value of the parameter
+    ///   - allocator: The allocator to use for the parameter
+    public func add(name: String, value: String?, allocator: Allocator = defaultAllocator) throws {
+        guard let value = value else {
+            return
+        }
```

Comment:
because all endpoint params are nullable. They are discarded here because C APIs don't have nullable concept (yet).

Reply:
I think it might be better to make value non nullable and the SDK only calls this functions when it has both name & value. Currently, this implementation makes it look like that we can successfully just add a name with no value which isn’t true. However, I am not sure how this fits into the big picture and if it is worth the effort to make this change.

Reply:
By the endpoints spec, params are nullable hence C must allow taking nullable values. This is safe guard for C code than SDK code.

Reply:
Also, this is by luck (actually rules are not exercising that path) here both `nil` and `not present` are working in C. @DmitriyMusatkin can share more information here.

Reply:
so to provide some context, this is basically a question about where crt draws a line for interface.

for ep2, a set of all possible parameters is defined for each service. some of those are required, some are optional. 

so basically, we are guaranteed that param name is never null (because we know all possible params at codegen time), but depending on type of request some parameter values might not be present (ex. bucket name on list buckets request). 

from sdk perspective set of parameters is basically constant, but some have null values. 

CRT currently asks to pass only the parameters that have concrete values (i.e. non null). This has some implementation benefits, but from interface perspective that means that caller needs to filter out all nulls. This does not affect correctness of eval in any way, and if CRT was to take nulls for param values, we would either make it no-op at the time of context construction or add null checks throughout eval (which i hate). 

anyways, discussed it with mike last week and we feel like its better design to not allow nullable types here. should we have a quick meeting next week, when mike is back to close out on this?

Reply:
Thanks, that makes sense.